### PR TITLE
Wip aj443 wind core interface

### PR DESCRIPTION
**Pull request by gameboo**

### CODE REVIEW

File: builds/RV64ACDFIMSUxCHERI_Toooba_bluesim/Makefile [hunk 1]
```diff
@@ -11,6 +11,23 @@ EXTRA_DIRS = $(RISCY_HOME)/../../src_Verifier:$(RISCY_HOME)/../../src_Verifier/B
 # ================================================================
 # RISC-V config macros passed into Bluespec 'bsc' compiler
 
+BSC_COMPILATION_FLAGS += \
```
Comment (PeterRugg) [inline builds/RV64ACDFIMSUxCHERI_Toooba_bluesim/Makefile:14]: Where did these come from? How did we live without them before?

File: src_Core/Core/CoreW.bsv [hunk 2]
```diff
@@ -110,9 +119,37 @@ import DM_CPU_Req_Rsp ::*;
 // ================================================================
 // The Core module
 
-(* synthesize *)
-module mkCoreW #(Reset dm_power_on_reset)
-               (CoreW_IFC #(N_External_Interrupt_Sources));
+typedef WindCoreMid #( // AXI lite subordinate control port parameters
+                       21, 32, 0, 0, 0, 0, 0
```
Comment (PeterRugg) [inline src_Core/Core/CoreW.bsv:123]: What are 21 and 32 here?

File: src_Core/Core/CoreW.bsv [hunk 3]
```diff
@@ -172,14 +209,16 @@ module mkCoreW #(Reset dm_power_on_reset)
...
+       prepend_AXI4_Master_id (0, zero_AXI4_Master_user (proc.master1));
    // Bridge for uncached expernal bus transactions.
    let uncached_mem_shim <- mkAXI4ShimFF(reset_by all_harts_reset);
 
    // handle cached interface
    // AXI4 tagController
-   TagControllerAXI#(Wd_MId, Wd_Addr, Wd_Data) tagController <- mkTagControllerAXI(reset_by all_harts_reset); // TODO double check if reseting like this is good enough
-   mkConnection(proc.master0, tagController.slave, reset_by all_harts_reset);
+   TagControllerAXI #(Wd_MId, Wd_Addr, Wd_Data)
+     tagController <- mkTagControllerAXI (reset_by all_harts_reset); // TODO double check if reseting like this is good enough
```
Comment (PeterRugg) [inline src_Core/Core/CoreW.bsv:220]: I guess the TODO got resolved?

File: src_Testbench/SoC/SoC_Top.bsv [hunk 4]
```diff
@@ -90,15 +91,9 @@ import Debug_Module     :: *;
...
-   // Set core's verbosity
-   method Action  set_verbosity (Bit #(4)  verbosity, Bit #(64)  logdelay);
 
 `ifdef INCLUDE_GDB_CONTROL
-   // DMI (Debug Module Interface) facing remote debugger
-   interface DMI dmi;
-
-   // Non-Debug-Module Reset (reset all except DM)
-   interface Client #(Bool, Bool) ndm_reset_client;
+  interface AXI4Lite_Slave #(7, 32, 0, 0, 0, 0, 0) debug_subordinate;
```
Comment (PeterRugg) [inline src_Testbench/SoC/SoC_Top.bsv:96]: What are 7 and 32 here?

**Review by PeterRugg — APPROVED** [review]
Looks fine to me

File: src_Core/Core/CoreW.bsv [hunk 5]
```diff
@@ -215,46 +254,61 @@ module mkCoreW #(Reset dm_power_on_reset)
    TV_Encode_IFC tv_encode <- mkTV_Encode;
 `endif
 
+   function do_release (restartRunning) = action
+      plic.set_addr_map (zeroExtend (soc_map.m_plic_addr_range.base),
+                         zeroExtend (rangeTop(soc_map.m_plic_addr_range)));
+      proc.start (restartRunning, soc_map_struct.pc_reset_value, 0, 0);
+      //proc.set_verbosity (verbosity);
```
Comment (francislaus) [inline src_Core/Core/CoreW.bsv:261]: Remove if not likely needed in the future.

File: src_Core/Core/CoreW.bsv [hunk 6]
```diff
@@ -369,30 +423,37 @@ module mkCoreW #(Reset dm_power_on_reset)
...
+   // ================================================================
+   // Connect the local bus
+
+   // Masters on the local bus
+   Vector #( CoreW_Bus_Num_Masters
+           , AXI4_Master #( Wd_CoreW_Bus_MId, Wd_Addr, Wd_Data
+                          , Wd_AW_User, Wd_W_User, Wd_B_User
+                          , Wd_AR_User, Wd_R_User))
+      master_vector = newVector;
    //let master_vector = newVector;
```
Comment (francislaus) [inline src_Core/Core/CoreW.bsv:438]: Remove?

File: src_Core/Core/CoreW.bsv [hunk 7]
```diff
@@ -369,30 +423,37 @@ module mkCoreW #(Reset dm_power_on_reset)
...
-   Vector#(Num_Slaves_2x3, AXI4_Slave #(Wd_SId_2x3, Wd_Addr, Wd_Data,
-                                        Wd_AW_User, Wd_W_User, Wd_B_User,
-                                        Wd_AR_User, Wd_R_User))
-                                        slave_vector = newVector;
+   Vector #( CoreW_Bus_Num_Slaves
+           , AXI4_Slave #( Wd_CoreW_Bus_SId, Wd_Addr, Wd_Data
+                         , Wd_AW_User, Wd_W_User, Wd_B_User
+                         , Wd_AR_User, Wd_R_User))
+      slave_vector = newVector;
    //let slave_vector = newVector;
```
Comment (francislaus) [inline src_Core/Core/CoreW.bsv:450]: Remove?

File: src_Core/Core/CoreW.bsv [hunk 8]
```diff
@@ -423,78 +484,118 @@ module mkCoreW #(Reset dm_power_on_reset)
...
-
-   method Action start (Bool is_running, Bit #(64) tohost_addr, Bit #(64) fromhost_addr);
-      plic.set_addr_map (zeroExtend (soc_map.m_plic_addr_range.base),
-                         zeroExtend (rangeTop(soc_map.m_plic_addr_range)));
-
-      let pc = soc_map_struct.pc_reset_value;
-      proc.start (is_running, pc, tohost_addr, fromhost_addr);
-
-`ifdef INCLUDE_GDB_CONTROL
-      // Save for potential future use by rl_dm_harts_reset
```
Comment (francislaus) [inline src_Core/Core/CoreW.bsv:447]: Did we lose this functionality? If so, is there no need for this anymore?

File: src_Core/Core/CoreW.bsv [hunk 9]
```diff
@@ -423,78 +484,118 @@ module mkCoreW #(Reset dm_power_on_reset)
...
 
-      $display ("%0d: %m.method start: proc.start (pc %0h, tohostAddr %0h, fromhostAddr %0h)",
-                cur_cycle, pc, tohost_addr, fromhost_addr);
-   endmethod
+   let fromDbgReset <- mkPulseWire (reset_by porReset);
+   Reg #(UInt #(8)) ndm_reset_delay <- mkReg (0, reset_by porReset);
+   Reg #(Bool) ndm_reset_restart_running <- mkReg (False, reset_by porReset);
+   rule rl_debug_module_send_reset (ndm_reset_delay == 0);
+      let restartRunning <- debug_module.ndm_reset_client.request.get;
+      ndm_reset_delay <= 110;
```
Comment (francislaus) [inline src_Core/Core/CoreW.bsv:511]: Where is 110 coming from as reset delay?

File: src_Core/Core/CoreW.bsv [hunk 10]
```diff
@@ -423,78 +484,118 @@ module mkCoreW #(Reset dm_power_on_reset)
...
+   end
 
-   // Cached master to Fabric master interface
-   interface cpu_imem_master = tagController.master;
+   let nmirq_reg <- mkReg (False);
+   let nmirq_ifc = interface Put;
+      method put = writeReg (nmirq_reg);
+   endinterface;
+   rule rl_connect_nmirq;
+      // TODO: fixup; passing const False for now
```
Comment (francislaus) [inline src_Core/Core/CoreW.bsv:544]: Remove?

**Review by francislaus — APPROVED** [review]

File: src_Core/RISCY_OOO/procs/lib/Fpu.bsv [hunk 11]
```diff
@@ -95,7 +95,8 @@ module mkDoubleDiv(Server#(Tuple3#(Double, Double, FpuRoundMode), Tuple2#(Double
 `ifdef USE_XILINX_FPU
     let fpu <- mkXilinxFpDiv;
 `else
-    let int_div <- mkNonPipelinedDivider(3); // [sizhuo] size in RVFpu: 2
```
[the remark this answers is not in the document]
Comment (gameboo) [reply src_Core/RISCY_OOO/procs/lib/Fpu.bsv:98]: I think this is a rebase artifact... I had some local change in my branch before this got in CHERi or something along these lines...? @jonwoodruff does this ring a bell? Either way we agree on the code. Do we care about this comment?

File: src_Testbench/SoC/SoC_Top.bsv (hunk 4 above)
Comment (gameboo) [reply src_Testbench/SoC/SoC_Top.bsv:96]: they are is the address width and the data width of the dmi interface (at least they should be)

File: src_Core/Core/CoreW.bsv (hunk 2 above)
Comment (gameboo) [reply src_Core/Core/CoreW.bsv:123]: address width and data width into the axi4 lite "management" slave

File: src_Core/Core/CoreW.bsv (hunk 5 above)
Comment (gameboo) [reply src_Core/Core/CoreW.bsv:261]: can do :) thx

File: src_Core/Core/CoreW.bsv (hunk 6 above)
Comment (gameboo) [reply src_Core/Core/CoreW.bsv:438]: yup :)

File: src_Core/Core/CoreW.bsv (hunk 7 above)
Comment (gameboo) [reply src_Core/Core/CoreW.bsv:450]: yup yup :)

File: src_Core/Core/CoreW.bsv (hunk 8 above)
Comment (gameboo) [reply src_Core/Core/CoreW.bsv:447]: There is just a reset now, not a start method. and reseting, well.. restarts... (at least it should if there are no bugs :p)

File: src_Core/Core/CoreW.bsv (hunk 9 above)
Comment (gameboo) [reply src_Core/Core/CoreW.bsv:511]: from the P3 wrapper.

File: src_Core/Core/CoreW.bsv (hunk 10 above)
Comment (gameboo) [reply src_Core/Core/CoreW.bsv:544]: yup

**Pull request merged**